color/src/color.rs

```diff
@@ -0,0 +1,531 @@
+        let scale = if alpha == 0.0 { 1.0 } else { 1.0 / alpha };
+        AlphaColor::new(add_alpha(T::LAYOUT.scale(multiplied, scale), alpha))
+    }
+
+    /// Interpolate colors.
+    ///
+    /// Note: this function doesn't fix up hue in cylindrical spaces. It is
+    /// still be useful if the hue angles are compatible, particularly if
+    /// the fixup has been applied.
+    #[must_use]
+    pub fn lerp_rect(self, other: Self, t: f32) -> Self {
+        self + t * (other - self)
+    }
+
+    /// Apply hue fixup for interpolation.
+    ///
+    /// Adjust the hue angle of `other` so that linear interpolation results in
+    /// the expected hue direction.
+    pub fn fixup_hues(self, other: &mut Self, direction: HueDirection) {
+        if let Some(ix) = T::LAYOUT.hue_channel() {
+            fixup_hue(self.components[ix], &mut other.components[ix], direction);
+        }
+    }
+
+    /// Linearly interpolate colors, with hue fixup if needed.
+    #[must_use]
+    pub fn lerp(self, mut other: Self, t: f32, direction: HueDirection) -> Self {
+        self.fixup_hues(&mut other, direction);
+        self.lerp_rect(other, t)
+    }
+
+    #[must_use]
+    pub const fn mul_alpha(self, rhs: f32) -> Self {
+        let (multiplied, alpha) = split_alpha(self.components);
+        Self::new(add_alpha(T::LAYOUT.scale(multiplied, rhs), alpha * rhs))
+    }
+
+    /// Difference between two colors by Euclidean metric.
+    #[must_use]
+    pub fn difference(self, other: Self) -> f32 {
+        let x = self.components;
+        let y = other.components;
+        let (d0, d1, d2, d3) = (x[0] - y[0], x[1] - y[1], x[2] - y[2], x[3] - y[3]);
+        (d0 * d0 + d1 * d1 + d2 * d2 + d3 * d3).sqrt()
+    }
+}
+
+// Lossless conversion traits.
+
+impl<T: Colorspace> From<OpaqueColor<T>> for AlphaColor<T> {
+    fn from(value: OpaqueColor<T>) -> Self {
+        value.with_alpha(1.0)
+    }
+}
+
+impl<T: Colorspace> From<OpaqueColor<T>> for PremulColor<T> {
+    fn from(value: OpaqueColor<T>) -> Self {
+        Self::new(add_alpha(value.components, 1.0))
+    }
+}
+
+// Arithmetic traits. A major motivation for providing these is to enable
+// weighted sums, which are well defined when the weights sum to 1. In
+// addition, multiplication by alpha is well defined for premultiplied
+// colors in rectangular colorspaces.
+
+impl<T: Colorspace> core::ops::Mul<f32> for OpaqueColor<T> {
+    type Output = Self;
+
+    fn mul(self, rhs: f32) -> Self {
+        Self::new(self.components.map(|x| x * rhs))
+    }
+}
+
+impl<T: Colorspace> core::ops::Mul<OpaqueColor<T>> for f32 {
+    type Output = OpaqueColor<T>;
+
+    fn mul(self, rhs: OpaqueColor<T>) -> Self::Output {
+        rhs * self
+    }
+}
+
+impl<T: Colorspace> core::ops::Div<f32> for OpaqueColor<T> {
+    type Output = Self;
+
+    #[expect(
+        clippy::suspicious_arithmetic_impl,
+        reason = "somebody please teach clippy about multiplicative inverses"
```

Comment:
Is it worth linking to https://github.com/rust-lang/rust-clippy/issues/13652 here?